File: src/tools/finance/markov-distribution/volume-regime.ts
/** Mirrors `research/models/markov/volume_regime.py`. */

export type VolumeEnvironment = 'low_volume' | 'normal_volume' | 'high_volume';
export type VolumeRegime = 'low' | 'normal' | 'high';

export interface VolumeRegimeDetectionResult {
  environment_sequence: VolumeEnvironment[];
  current_environment: VolumeEnvironment;
}

function median(values: number[]): number {
  if (values.some(Number.isNaN)) return Number.NaN;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? (sorted[mid - 1] + sorted[mid]) / 2 : sorted[mid];
}

export function classifyVolumeRegime(
  volumes: number[],
  lookback = 20,
  thresholdMultiplier = 1.5,
): VolumeRegime[] {
  if (thresholdMultiplier <= 0) {
    throw new RangeError(`threshold_multiplier must be positive, got ${thresholdMultiplier}`);
  }
  if (lookback < 2) {
    throw new RangeError(`lookback must be >= 2, got ${lookback}`);
  }

  const n = volumes.length;
  if (n === 0) return [];

  const result: VolumeRegime[] = Array(n).fill('normal');
  if (n <= lookback) return result;

  for (let i = lookback; i < n; i++) {
    const rollingMedian = median(volumes.slice(i - lookback, i));
    if (rollingMedian <= 0 || Number.isNaN(rollingMedian)) continue;

    const value = volumes[i];
    if (Number.isNaN(value)) continue;
    if (value > thresholdMultiplier * rollingMedian) {
      result[i] = 'high';
    } else if (value < rollingMedian / thresholdMultiplier) {
      result[i] = 'low';
    }
  }

  return result;
}

export class VolumeRegimeDetector {
  private isFitted = false;

  constructor(
    public readonly lookback = 20,
    public readonly thresholdMultiplier = 1.5,
  ) {}

  get fitted(): boolean {
    return this.isFitted;
  }

  fit(volumes: number[]): VolumeRegimeDetectionResult {
    const labels = classifyVolumeRegime(volumes, this.lookback, this.thresholdMultiplier);
    const envMap: Record<VolumeRegime, VolumeEnvironment> = {
      high: 'high_volume',
      normal: 'normal_volume',
      low: 'low_volume',
    };
    const environmentSequence = labels.map(label => envMap[label]);

    this.isFitted = true;
    return {
      environment_sequence: environmentSequence,
      current_environment: environmentSequence.at(-1) ?? 'normal_volume',
    };
  }
}

File: research/models/markov/volume_regime.py
"""Volume regime detection mirroring ``src/tools/finance/markov-distribution/volume-regime.ts``."""

from __future__ import annotations

from typing import Literal

import numpy as np

VolumeRegime = Literal["low", "normal", "high"]
VolumeEnvironment = Literal["low_volume", "normal_volume", "high_volume"]


def classify_volume_regime(
    volumes: np.ndarray | list[float],
    lookback: int = 20,
    threshold_multiplier: float = 1.5,
) -> list[VolumeRegime]:
    """Classify volume as high, normal, or low relative to the previous rolling median."""
    if threshold_multiplier <= 0:
        raise ValueError(f"threshold_multiplier must be positive, got {threshold_multiplier}")

    arr = np.asarray(volumes, dtype=float)
    n = len(arr)
    if n == 0:
        return []
    if lookback < 2:
        raise ValueError(f"lookback must be >= 2, got {lookback}")

    result: list[VolumeRegime] = ["normal"] * n
    if n <= lookback:
        return result

    for i in range(lookback, n):
        window = arr[i - lookback : i]
        median = float(np.median(window))
        if median <= 0 or np.isnan(median):
            continue
        value = float(arr[i])
        if np.isnan(value):
            continue
        if value > threshold_multiplier * median:
            result[i] = "high"
        elif value < median / threshold_multiplier:
            result[i] = "low"

    return result


class VolumeRegimeDetector:
    """Map rolling-median volume labels to volume environment names."""

    def __init__(self, lookback: int = 20, threshold_multiplier: float = 1.5) -> None:
        self.lookback = lookback
        self.threshold_multiplier = threshold_multiplier
        self._fitted = False

    @property
    def fitted(self) -> bool:
        return self._fitted

    def fit(self, volumes: np.ndarray | list[float]) -> dict[str, list[VolumeEnvironment] | VolumeEnvironment]:
        labels = classify_volume_regime(volumes, lookback=self.lookback, threshold_multiplier=self.threshold_multiplier)
        env_map: dict[VolumeRegime, VolumeEnvironment] = {
            "high": "high_volume",
            "normal": "normal_volume",
            "low": "low_volume",
        }
        env_seq = [env_map[label] for label in labels]

        self._fitted = True
        return {
            "environment_sequence": env_seq,
            "current_environment": env_seq[-1] if env_seq else "normal_volume",
        }
